File: CodingSamples/DotNet/Fundamentals/Inheritance/ClassTypeTest/DemoApp2/Interval.cs
//primary constructor
class Interval(int min, int sec)
{
    //read-only properties (no set block)
    public int Minutes { get; } = min + sec / 60;

    public int Seconds { get; } = sec % 60;

    public int Time()
    {
        return 60 * Minutes + Seconds;
    }

    //overloading addition(+) operator
    public static Interval operator+(Interval lhs, Interval rhs)
    {
        return new Interval(lhs.Minutes + rhs.Minutes, lhs.Seconds + rhs.Seconds);
    }

    //overriding corresponding method of System.Object to return
    //a string containing values of fields in this instance
    public override string ToString()
    {
        if(Seconds >= 10)
            return Minutes + ":" + Seconds;
        return Minutes + ":0" + Seconds;
    }

    //overriding corresponding method of System object to return
    //an integer obtained by combining values of fields in this
    //instance such that the same integer is returned for two
    //instances with matching field values
    public override int GetHashCode()
    {
        return Minutes + Seconds;
    }

    //overriding corresponding method of System object to return
    //true or false depending on whether or not the values of fields
    //in this instance match with those in instance referred by 
    //the argument passed 
    public override bool Equals(object other)
    {
        if(other is Interval)
        {
            Interval that = (Interval)other; //explicit down-casting
            return this.Minutes == that.Minutes && this.Seconds == that.Seconds;
        }
        return false;
    }
}

File: CodingSamples/Java/Fundamentals/Generics/TypeParamTest/demo2/Interval.java
class Interval implements Comparable<Interval> {
    
    private int min;
    private int sec;

    public Interval(int m, int s) {
        min = m + s / 60;
        sec = s % 60;
    }

    public int minutes() {
        return min;
    }

    public int seconds() {
        return sec;
    }

    public int time() {
        return 60 * min + sec;
    }

    public Interval add(Interval other) {
        return new Interval(min + other.min, sec + other.sec);
    }

    //overriding toString() method of java.lang.Object to return
    //a string containing values of fields within the instance
    public String toString() {
        if(sec < 10)
            return min + ":0" + sec;
        return min + ":" + sec;
    }

    //overriding hashCode() method of java.lang.Object to return
    //an integer so that same integer is returned for two instances
    //with matching values of instance fields
    public int hashCode() {
        return min + sec;
    }

    public boolean equals(Object other) {
        if(other instanceof Interval){
            Interval that = (Interval) other; //explicit down-casting
            return min == that.min && sec == that.sec; 
        }
        return false;
    }

    public int compareTo(Interval that) {
        return this.time() - that.time();
    }
}
